### PR TITLE
MU plugin for the Server-Timing API

### LINKED ISSUE #3: Create a mini-MU plugin for the Server-Timing API
- Based on the [Performance Lab plugin implementation](https://github.com/WordPress/performance/tree/trunk/server-timing) create MU plugin for Server-Timing API

### Acceptance Criteria

- Add a new folder called `performance` to the `wordpress-develop/tests` directory.
- Create a new folder called `server-timing` and place the necessary files in it.
- Add new step in `performance.yml` action that add MU plugin in docker environment.
```
name: Install MU plugin
run: |
  mkdir ./src/wp-content/mu-plugins
  cp ./tests/performance/server-timing/mu-plugin.php ./src/wp-content/mu-plugins/mu-plugin.php
```

The MU Plugin should record and report the following metrics via the Server-Timing API:

- `wp-before-template`: Time it takes from WordPress core's `timer_start()` call until the end of `template_redirect` action
- `wp-template`: Time it takes from end of `template_redirect` action to start of `shutdown`
- `wp-total`: Sum of both of the above metrics

Note that in order to measure `wp-template` and output it in a header, we need to use an output buffer. That's okay since it'll be running on all tests, thus not skewing data.

[This Gist](https://gist.github.com/felixarntz/de5c697a1a16c2b892634b70216eb6c7) can serve as potential inspiration for the implementation of the MU plugin.

### PR DESCRIPTION
**Pull request by mukeshpanchal27**
The PR adds a MU plugin for Server-Timing API and a GH action step that add mu plugin in docker container.

Closes #3

### CODE REVIEW

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (felixarntz) [inline tests/performance/plugins/server-timing/mu-plugin.php]: This is slightly off, as right now it measures the total time. It needs to only measure the time starting from the above `template_redirect` hook.

You can define a variable like `$template_start` in that hook callback that stores the current `microtime( true )` and then use that in here instead of the `$timestart` global, which is only needed for the `before-template` metric.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (felixarntz) [inline tests/performance/plugins/server-timing/mu-plugin.php]: Nit-pick, but I'd advise to start the output buffer at the end of this, right here.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (felixarntz) [inline tests/performance/plugins/server-timing/mu-plugin.php]: See above nit-pick, maybe best to end the output buffer right away here.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (felixarntz) [inline tests/performance/plugins/server-timing/mu-plugin.php]: This shouldn't be needed strictly speaking, since the variable is also ensured to be an array above. If it wasn't an array here, it also would be missing the `before-template` key which is unconditionally checked further below.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (felixarntz) [inline tests/performance/plugins/server-timing/mu-plugin.php]: This hook should rather happen as late as possible, since we only want to start the output buffer as close as possible to the template actually being rendered. For example, other hook callbacks potentially added should run before this one.

So I would suggest to change this to `9999`.

**Review by felixarntz — COMMENTED** [review]
@mukeshpanchal27 Looks like this includes the code from #6 and will need to be refreshed once that PR is merged.

But I already took a look at the new code in this PR and left some feedback. Looks almost good to go!

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (joemcgill) [reply tests/performance/plugins/server-timing/mu-plugin.php]: Perhaps `PHP_INT_MAX` instead of `9999`?

**Comment by joemcgill** [comment]
> Looks like this includes the code from https://github.com/10up/wordpress-develop/pull/6

To make reviews easier, you could set the target branch to the [setup-initial-performance-action](https://github.com/10up/wordpress-develop/tree/setup-initial-performance-action) branch and mark this a blocked until after #6 is merged. That way no refactor will be needed and this PR should automatically be updated to point to the feature branch once that one gets merged.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (mukeshpanchal27) [reply tests/performance/plugins/server-timing/mu-plugin.php]: Thanks for the review and pointing it out. I have address this in [b2581a7](https://github.com/10up/wordpress-develop/pull/7/commits/b2581a70662d9917e23094359c0e2ce47dcd467d)

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (mukeshpanchal27) [reply tests/performance/plugins/server-timing/mu-plugin.php]: Yes. Make sense.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (mukeshpanchal27) [reply tests/performance/plugins/server-timing/mu-plugin.php]: Thanks for the feedback, I have updated priority for `template_redirect` to `PHP_INT_MAX`

**Review by joemcgill — APPROVED** [review]
The updates to the mu-plugin look good to me.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (felixarntz) [inline tests/performance/plugins/server-timing/mu-plugin.php]: Nit-pick: Better not to make `$template_start` a global variable, as it's only needed in here and the closure below. You can pass it to the other closure via `use()`.

Maybe we should even keep `$server_timing_values` as just a local variable for now. At least for now, the way this code is written there is no real benefit in having it as a global. That would change if we want to add server timing values from elsewhere, but at least now we don't, so maybe best to avoid the global.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (felixarntz) [inline tests/performance/plugins/server-timing/mu-plugin.php]: See above, here you could access the local variables instead of using globals, e.g. like:
```suggestion
			function() use ( $server_timing_values, $template_start ) {
```

**Review by felixarntz — COMMENTED** [review]
@mukeshpanchal27 One small point of feedback on avoiding globals where not necessary. That in itself is not necessarily a blocker, but now that #6 was merged, we need to also refresh this branch and resolve the merge conflict, so maybe you can address my new feedback then as well.

Related is @joemcgill's comment from https://github.com/10up/wordpress-develop/pull/7#issuecomment-1412754388, that way we can avoid such conflicts in the future.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (mukeshpanchal27) [reply tests/performance/plugins/server-timing/mu-plugin.php]: Thanks @felixarntz Good catch.

File: tests/performance/plugins/server-timing/mu-plugin.php
Comment (mukeshpanchal27) [reply tests/performance/plugins/server-timing/mu-plugin.php]: Thanks for the suggestion.

**Comment by mukeshpanchal27** [comment]
Thanks for the feedback. @joemcgill I have addressed your feedback, which you have asked for [here](https://github.com/10up/wordpress-develop/pull/6#pullrequestreview-1281165914).

**Comment by joemcgill** [comment]
Thanks @mukeshpanchal27. The updates look good to me.

File: .github/workflows/performance.yml
Comment (felixarntz) [inline .github/workflows/performance.yml]: Nit-pick: It would make more sense to me to have this file in `./tests/performance/mu-plugins/server-timing.php`, as that would indicate better where it should go. Seems unnecessary to use a different structure in the source directory for this MU plugin.

**Review by felixarntz — APPROVED** [review]
@mukeshpanchal27 Thank you for making the iterations, LGTM. Tiny feedback below, but not blocking!

File: .github/workflows/performance.yml
Comment (mukeshpanchal27) [reply .github/workflows/performance.yml]: Thanks @felixarntz for the review. https://github.com/10up/wordpress-develop/pull/7/commits/85a5c96fe2bbccc806c4b8eb06863a26c4285617 updates the MU plugin path.

**Pull request merged**
